### PR TITLE
Add support for docker hosted repository

### LINKED ISSUE #4: Add repository docker(hosted) support

### PR DESCRIPTION
**Pull request by akaRem**
Closes #4

### CODE REVIEW

File: nexus3/client/repository_docker_hosted.go [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+	CleanupPolicies []string
+}
+
+func repositoryDockerHostedToRPCObj(v *RepositoryDockerHosted) interface{} {
+	output := map[string]interface{}{
+		// Field reuired to create, ignored otherwise
+		"recipe": "docker-hosted",
+		"name":   v.Name,
+		"online": v.Enabled,
+		// These fields required and may be "" 🤷‍♂️
```
Comment (akaRem) [inline nexus3/client/repository_docker_hosted.go:39]: These fields are required but allowed to be ""

File: nexus3/client/repository_docker_hosted.go [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+	Docker  struct {
+		HTTPPort          int
+		HTTPSPort         int
+		ForceAuth         bool
+		V1ProtocolEnabled bool
+	}
+	Storage struct {
+		BlobStoreName             string
+		MIMETypeValidationEnabled bool
+		// TODO: There should be better names for constants
```
Comment (akaRem) [inline nexus3/client/repository_docker_hosted.go:24]: Reformat note

File: nexus3/client/repository_docker_hosted.go [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+				"v1Enabled":      v.Docker.V1ProtocolEnabled,
+			},
+			"storage": map[string]interface{}{
+				"blobStoreName":               v.Storage.BlobStoreName,
+				"strictContentTypeValidation": v.Storage.MIMETypeValidationEnabled,
+				"writePolicy":                 v.Storage.WritePolicy,
+				"latestPolicy":                false,
+			},
+			"cleanup": map[string]interface{}{
+				// FIXME: Nexus gives 0Fs about existance of mentioned polici names,
```
Comment (akaRem) [inline nexus3/client/repository_docker_hosted.go:57]: typo in `polici`

File: nexus3/client/repository_docker_hosted.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+				// So we need to verify that they exist before we request a new repo
+				// with these policies used
+				"policyName": v.CleanupPolicies,
+			},
+		},
+	}
+	if v.Storage.WritePolicy == "ALLOW_LATEST_TAG" {
+		output["attributes"].(map[string]interface{})["storage"].(map[string]interface{})["writePolicy"] = "ALLOW_ONCE"
```
Comment (akaRem) [inline nexus3/client/repository_docker_hosted.go:65]: 😱

File: nexus3/client/repository_docker_hosted.go [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+			RedeployLatestEnabled     bool
+		}{
+			BlobStoreName:             d.Get("attributes.storage.blobStoreName").String(),
+			MIMETypeValidationEnabled: d.Get("attributes.storage.strictContentTypeValidation").Bool(),
+			WritePolicy:               d.Get("attributes.storage.writePolicy").String(),
+		},
+	}
+
+	if d.Get("attributes.storage.latestPolicy").Bool() {
+		output.Storage.WritePolicy = "ALLOW_LATEST_TAG"
```
Comment (akaRem) [inline nexus3/client/repository_docker_hosted.go:100]: Do we want constants?

File: nexus3/client/repository_docker_hosted.go [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+
+	output, err := repositoryDockerFromGJSON(*gjData)
+	if err != nil {
+		return nil, errors.Wrapf(err, "can't construct Cleanup policy from JSON %s", gjData.Raw)
+	}
+
+	return output, nil
+}
+
+// RepositoryDockerHostedRead reads  hosted docker repository
```
Comment (akaRem) [inline nexus3/client/repository_docker_hosted.go:126]: extra space

File: nexus3/client/repository_docker_hosted_test.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+		Storage: struct {
+			BlobStoreName             string
+			MIMETypeValidationEnabled bool
+			WritePolicy               string
+			RedeployLatestEnabled     bool
+		}{
+			BlobStoreName: "default",
+			WritePolicy:   "ALLOW_LATEST_TAG",
+		},
+		// CleanupPolicies: []string{"aaa"},
```
Comment (akaRem) [inline nexus3/client/repository_docker_hosted_test.go:80]: remove

File: nexus3/client/repository_docker_hosted.go (hunk 4 above)
Comment (kuzmacska) [reply nexus3/client/repository_docker_hosted.go:68]: omg

File: nexus3/client/repository_docker_hosted.go (hunk 5 above)
Comment (kuzmacska) [reply nexus3/client/repository_docker_hosted.go:103]: not here imho

**Pull request merged**